tests/virtual_device.rs

```diff
@@ -0,0 +1,56 @@
+//use evdev::{Device};
+use std::error::Error;
+use std::thread::sleep;
+use std::time::Duration;
+
+use tokio_1 as tokio;
+use tokio::time::timeout;
+
+use evdev::{uinput::VirtualDeviceBuilder, AttributeSet, EventType, InputEvent, Key};
+
+#[tokio::test]
+async fn test_virtual_device_actually_emits() -> Result<(), Box<dyn Error>> {
+    let mut keys = AttributeSet::<Key>::new();
+    let virtual_device_name = "fake-keyboard";
+    keys.insert(Key::KEY_ESC);
+
+    let mut device = VirtualDeviceBuilder::new()?
+        .name(virtual_device_name)
+        .with_keys(&keys)?
+        .build()
+        .unwrap();
+
+    let mut maybe_device = None;
+    sleep(Duration::from_millis(500));
+    for (_i, d) in evdev::enumerate() {
+        println!("{:?}", d.name());
+        if d.name() == Some(virtual_device_name) {
+            maybe_device = Some(d);
+            break
+        }
+    }
+    assert_eq!(maybe_device.is_some(), true);
+    let listen_device = maybe_device.unwrap();
+
+    let type_ = EventType::KEY;
+    let code = Key::BTN_DPAD_UP.code();
```

Comment:
d'oh! This key isn't enabled on the virtual device, changing it to `KEY_ESC` makes this test pass. But that still leaves me with the problem of why input events sent to the virtual device in my keyboard remapper don't seem to make it to userspace...

Anyway, if y'all are interested in an extra test case I'd be happy to move this into `src/uinput.rs`